### DISCUSSION TITLE
How can I create a connection even if my RabbitMQ Server is offline?

### DISCUSSION THREAD

**Discussion by marco-carvalho** [rabbitmq/rabbitmq-dotnet-client#1513] [General]
### Describe the bug

When deploying a C# program that establishes a `sharedConnection` to a RabbitMQ server with autorecovery enabled, the connection persists as expected even if the RabbitMQ server goes offline. However, if the server remains offline and the program is redeployed, attempting to re-establish `sharedConnection` results in a null reference, and subsequent connection attempts throw an exception instead of waiting to recover when the server becomes available again.

### Reproduction steps

1. Ensure the RabbitMQ server is running and accessible.
2. Deploy a C# application that connects to RabbitMQ and initializes a `sharedConnection` with autorecovery enabled.
3. Turn off the RabbitMQ server, simulating a server outage. Observe that `sharedConnection` still exists (is not null) due to autorecovery being on.
4. Redeploy the C# application while the RabbitMQ server is still offline.
5. Observe that `sharedConnection` is now null, and attempts to connect to RabbitMQ throw an exception.

### Expected behavior

Even when the RabbitMQ server is offline at the time of the application deployment, the client should initialize sharedConnection in a manner that allows autorecovery to attempt reconnections until the server becomes available again, instead of immediately throwing an exception.

**Comment by michaelklishin** [comment]
RabbitMQ clients can only [connect to a running node](https://www.rabbitmq.com/docs/connections). This is not different from the absolute majority of messaging and data services.

The only option for your applications is to retry and/or use a list of endpoints when connecting (assuming that at least one of them would be online).

Several related doc guides:

 * [Connecting to a list of endpoints](https://www.rabbitmq.com/docs/api-guide#endpoints-list) (most clients provide this feature)
 * [Connection Recovery](https://www.rabbitmq.com/docs/api-guide#recovery) (most clients provide this feature)
 * [Publisher side data safety](https://www.rabbitmq.com/docs/publishers#data-safety)
 * [Publisher confirms and consumer acknowledgements](https://www.rabbitmq.com/docs/confirms)

**Comment by michaelklishin** [comment]
Connection recovery is not the same scenarios as an initial connection in many ways. Initial connection with retries is usually trivial to implement in your own code:

 * Try to connect and catch relevant exceptions
 * In case of failure, retry (possibly with a delay)
 * After a certain number of attempts or a time period, applications should give up and quit, to avoid retrying forever

That's all there is to it.
